spec/java/util/HashMap.main.lsl

```diff
@@ -0,0 +1,642 @@
+        if (action MAP_HAS_KEY(this.storage, key))
+        {
+            val entry: Map_Entry<Object, Object> = action MAP_GET(this.storage, key);
+            result = action CALL_METHOD(entry, "getValue", [])
+            action MAP_REMOVE(this.storage, key);
+            this.modCount += 1;
+        }
+        else
+        {
+            result = null;
+        }
+    }
+
+
+    fun *.remove (@target self: HashMap, key: Object, value: Object): boolean
+    {
+        result = false;
+
+        if (action MAP_HAS_KEY(this.storage, key))
+        {
+            val entry: Map_Entry<Object, Object> = action MAP_GET(this.storage, key);
+            val curValue: Object = action CALL_METHOD(entry, "getValue", []);
+            // #question: It will throw NPE if curValue == null and value == null ? Or not ?
+            if (action OBJECT_EQUALS(curValue, value))
+            {
+                action MAP_REMOVE(this.storage, key);
+                this.modCount += 1;
+                result = true;
+            }
+        }
+    }
+
+
+    fun *.replace (@target self: HashMap, key: Object, value: Object): Object
+    {
+        result = null;
+        if (action MAP_HAS_KEY(this.storage, key))
+        {
+            val entry: Map_Entry<Object, Object> = action MAP_GET(this.storage, key);
+            action CALL_METHOD(entry, "setValue", [value]);
+        }
+    }
+
+
+    fun *.replace (@target self: HashMap, key: Object, oldValue: Object, newValue: Object): boolean
+    {
+        result = false;
+        if (action MAP_HAS_KEY(this.storage, key))
+        {
+            val entry: Map_Entry<Object, Object> = action MAP_GET(this.storage, key);
+            val curValue: Object = action CALL_METHOD(entry, "getValue", []);
+            if (action OBJECT_EQUALS(curValue, oldValue))
+            {
+                action CALL_METHOD(entry, "setValue", [newValue]);
+                result = true;
+            }
+        }
+    }
+
+
+    fun *.replaceAll (@target self: HashMap, function: BiFunction): void
+    {
+        if (function == null)
+            _throwNPE();
+
+        val thisSize: int = action MAP_SIZE(this.storage);
+        if (thisSize > 0)
+        {
+            val unseen: map<Object, Map_Entry<Object, Object>> = action MAP_CLONE(this.storage);
+            val expectedModCount: int = this.modCount;
+            var i: int = 0;
+            action LOOP_FOR(
+                i, 0, thisSize, +1,
+                replaceAll_loop(unseen, function)
+            );
+            _checkForComodification(expectedModCount);
+        }
+    }
+
+
+    @Phantom proc replaceAll_loop (unseen: map<Object, Map_Entry<Object, Object>>, function: BiFunction): void
+    {
+        val curKey: Object = action MAP_GET_ANY_KEY(unseen);
+        val entry: Map_Entry<Object, Object> = action MAP_GET(this.storage, curKey);
+        val curValue: Object = action CALL_METHOD(entry, "getValue", []);
+        action CALL(function, [curKey, curValue]);
```

Comment:
Return value from the user method is ignored.

Reply:
Fixed;